feat(appengine): Update helloworld to .NET 6

jahjahbin opened the pull request:
Migrate the `HelloWorld` GAE Flex sample to .NET 6.

Background: 
GAE Flex will soon support .NET 6. We'd like to update all the code samples for that, starting with `HelloWorld`. It looks like the basic code is much simpler since the last version, 3.1.

This barebones sample was created via `dotnet new web` and adapted to match what the old HelloWorld sample did, i.e. read a variable from configuration or env vars.

amanda-tarafa commented on appengine/flexible/HelloWorld/HelloWorld.sln:
Are you planning on adding a solution file for the whole of appengine/flexible applications? If so, it's fine to delete this one, but do create the new one on this PR and include this project there.
We require a solution file as it's easier for maintainers using Visual Studio to open everything through a single file.

jahjahbin replied:
There seems to already be a solution file present under `appengine/flexible/AppEngineFlex.sln`. It links to all of the Flex samples. Will that do?

amanda-tarafa commented on appengine/flexible/HelloWorld/Program.cs:
```diff
@@ -13,29 +13,9 @@
  * License for the specific language governing permissions and limitations under
  * the License.
  */
+var builder = WebApplication.CreateBuilder(args);
```
nit, blank line between the licence and the start of the code
```suggestion

var builder = WebApplication.CreateBuilder(args);

```

jahjahbin replied:
Done.

amanda-tarafa commented on appengine/flexible/HelloWorld/Startup.cs:
```diff
@@ -1,61 +0,0 @@
-
-namespace HelloWorld
-{
-    // [START gae_flex_quickstart]
```
Should you move this region tag to Program.cs now? How are you planning on referencing the sample from docs?

jahjahbin replied:
Good point, I moved it to Program.cs.

amanda-tarafa replied:
Yep, exactly.

amanda-tarafa commented on appengine/flexible/HelloWorld/Properties/launchSettings.json:
```diff
@@ -8,20 +8,21 @@
-      "commandName": "IISExpress",
+    "HelloWorld": {
+      "commandName": "Project",
+      "dotnetRunMessages": true,
```
nit: The only real change here seems to be the setting of the `dotnetRunMessages` field. The rest is just because the order of the progiles changed. Would you mind reverting the order change so we see the real change?

jahjahbin replied:
Done.

amanda-tarafa reviewed (changes requested):
Apart from the comments a couple othre things:

- Tests are failing. Currently they rely on Casper, but probably on a version of Casper that won't work with .NET 6 etc. I think it's better if tests are also [modernized](https://learn.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-6.0).
- You'll need to also update the README file within the HelloWorld project folder. It's fine if that just points to public AppEngine Flex docs, it is there for the (unlikely) user tha lands directly on GitHub.

jahjahbin commented:
> Apart from the comments a couple othre things:
> 
> * Tests are failing. Currently they rely on Casper, but probably on a version of Casper that won't work with .NET 6 etc. I think it's better if tests are also [modernized](https://learn.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-6.0).
> * You'll need to also update the README file within the HelloWorld project folder. It's fine if that just points to public AppEngine Flex docs, it is there for the (unlikely) user tha lands directly on GitHub.

Sounds good, I updated both and deleted the Casper tests. I followed the linked doc from Microsoft, so please let me know if I credited them property.

jinglundong commented on appengine/flexible/HelloWorld.Tests/FunctionalTests.cs:
```diff
@@ -0,0 +1,27 @@
+        var response = await client.GetAsync(url);
+        response.EnsureSuccessStatusCode();
+	var body = await response.Content.ReadAsStringAsync();
+	Assert.True(string.Equals("Hello World!", body), $"Unexpected response: {body}");
```
[optional] Does it make sense to also print and verify the major version of aspnetcore?

jahjahbin replied:
I think that makes more sense for app engine. This test isn't deploying anything to GAE, just spinning up a local server for testing.

jahjahbin commented:
Friendly ping @amanda-tarafa , could you take a look at the revised PR?

Sorry it took me a while to address the comments, I was oncall when I sent it originally.

amanda-tarafa commented on appengine/flexible/HelloWorld.Tests/FunctionalTests.cs:
```diff
@@ -0,0 +1,27 @@
+using Microsoft.AspNetCore.Mvc.Testing;
+using Xunit;
+
+// Provides simple tests to demonstrate calling the sample app. This is
+// based on the straightforward examples provided by Microsoft here:
+// https://learn.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-6.0
```
nit: Add this within `<summary><\summary>`

jahjahbin replied:
Done.

amanda-tarafa commented on appengine/flexible/HelloWorld.Tests/FunctionalTests.cs:
```diff
@@ -0,0 +1,27 @@
+        _factory = factory;
+    }
+
+    [Theory]
+    [InlineData("/")]
+    public async Task TestGetSucceeds(string url)
+    {
+        var client = _factory.CreateClient();
+        var response = await client.GetAsync(url);
```
If you are testing for a single case, which is totally fine, then there's no need to make this a Theory.
```suggestion
    [Fact]
    public async Task TestGetSucceeds()
    {
        var client = _factory.CreateClient();
        var response = await client.GetAsync("/");
```

jahjahbin replied:
Done.

amanda-tarafa commented on appengine/flexible/HelloWorld.Tests/FunctionalTests.cs:
```diff
@@ -0,0 +1,27 @@
+        var response = await client.GetAsync(url);
+        response.EnsureSuccessStatusCode();
+	var body = await response.Content.ReadAsStringAsync();
+	Assert.True(string.Equals("Hello World!", body), $"Unexpected response: {body}");
```
The message when this assert fails, includes the expected and the actual value, etc.
```suggestion
	Assert.Equal("Hello World!", body);
```

jahjahbin replied:
Done.

amanda-tarafa commented on appengine/flexible/HelloWorld/runTests.ps1:
Do not delete this file, this is the ones that is used for running tests. Just update it to look like this: https://github.com/GoogleCloudPlatform/dotnet-docs-samples/blob/main/kms/api/runTests.ps1

jahjahbin replied:
Done.

amanda-tarafa commented on appengine/flexible/HelloWorld.Tests/HelloWorld.Tests.csproj:
```diff
@@ -0,0 +1,29 @@
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
+    <PackageReference Include="coverlet.collector" Version="3.1.2">
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
```
On this repo we don't use coverlet, we use JUnitTestLogger, like [here ](https://github.com/GoogleCloudPlatform/dotnet-docs-samples/blob/25fd7d9940705b60635c2d07f3298f084a35656c/kms/api/Kms.Samples.Tests/Kms.Samples.Tests.csproj#L8)
```suggestion
    <PackageReference Include="JUnitTestLogger" Version="1.1.0" />
```

jahjahbin replied:
Done.

amanda-tarafa reviewed (commented):
Sorry for the delay, this fell off my radar.

This is looking good. I just left a couple of comments that will make tests run. (And yes, the top level solution will do, I didn't realize there was one).

amanda-tarafa reviewed (commented):
Tests are not being run yet, and that's because how the new folders are structured.

The runTests.ps1 file needs to be on the parent folder of the test project, or inside the test project itself. And in general we have a folder that contains both the sample project and the test project (since tests were with Casper before this didn't apply here). See [KMS](https://github.com/GoogleCloudPlatform/dotnet-docs-samples/tree/main/kms/api) as an example.

So, can I ask you to restructure as follows:

- Keep the HelloWorld folder.
- Inside the HelloWorld folder create a HelloWorld.Sample folder and move the project there.
- Inside the HelloWorkd folder create a HelloWorld.Tests folder and move the test project there.
- Keep the Readme, runTests.ps1 etc. on the HelloWorld folder.
- Update the solution.

Sorry for not realizing this before.

(And then follow this structure for the other samples when you update them)

amanda-tarafa commented on appengine/flexible/AppEngineFlex.sln:
```diff
@@ -23,12 +23,16 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Pubsub", "Pubsub\Pubsub.csp
 EndProject
 Project("{6EC3EE1D-3C4E-46DD-8F32-0CC8E7565705}") = "F#", "F#\F#.fsproj", "{C55A9BF5-F4FA-4E94-845C-302CBA878946}"
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "HelloWorld", "HelloWorld", "{B300174E-692E-4378-A6C4-8E6C365F815E}"
+EndProject
```
```suggestion
```

amanda-tarafa commented on appengine/flexible/HelloWorld/HelloWorld.Tests/FunctionalTests.cs:
```diff
@@ -0,0 +1,28 @@
+using Microsoft.AspNetCore.Mvc.Testing;
```
License missing

```suggestion
/*
 * Copyright (c) 2023 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

using Microsoft.AspNetCore.Mvc.Testing;
```

amanda-tarafa reviewed (commented):
Two tiny nits that I'll commit myself and then merge.

Thanks!

amanda-tarafa commented:
It's all good now. I'll approve and merge once the tests have run once last time, just to make sure that I didn't break anything with the nits I commit.

amanda-tarafa commented on appengine/flexible/AppEngineFlex.sln:
```diff
@@ -23,12 +23,14 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Pubsub", "Pubsub\Pubsub.csp
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "AntiForgery", "AntiForgery\AntiForgery.csproj", "{B7B77EE8-60C8-49C3-83D2-AD907B882332}"
 EndProject
 Project("{6EC3EE1D-3C4E-46DD-8F32-0CC8E7565705}") = "F#", "F#\F#.fsproj", "{C55A9BF5-F4FA-4E94-845C-302CBA878946}"
 EndProject
```
```suggestion
EndProject
Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "HelloWorld", "HelloWorld", "{B300174E-692E-4378-A6C4-8E6C365F815E}"
EndProject
```

jahjahbin commented:
Thanks!
